src/ethereum/prague/vm/eof/utils.py

```diff
@@ -0,0 +1,271 @@
+"""
+Utility functions for EOF containers.
+"""
+from ethereum.base_types import Uint
+
+from ..exceptions import InvalidEof
+from . import EOF_MAGIC, EofMetadata
+
+
+def metadata_from_container(
+    container: bytes,
+    validate: bool,
+    is_deploy_container: bool,
+    is_init_container: bool,
+) -> EofMetadata:
+    """
+    Validate the header of the EOF container.
+
+    Parameters
+    ----------
+    container : bytes
+        The EOF container to validate.
+    validate : bool
+        Whether to validate the EOF container. If the container is simply read
+        from an existing account, it is assumed to be validated. However, if
+        the container is being created, it should be validated first.
+    is_deploy_container : bool
+        Whether the container is a deploy container for EOFCREATE/ create
+        transactions.
+    is_init_container : bool
+        Whether the container is an init container for EOFCREATE/
+        create transactions.
+
+    Returns
+    -------
+    EofMetadata
+        The metadata of the EOF container.
+
+    Raises
+    ------
+    InvalidEof
+        If the header of the EOF container is invalid.
+    """
+    counter = len(EOF_MAGIC) + 1
+
+    # Get the one byte kind type
+    if validate and len(container) < counter + 1:
+        raise InvalidEof("Kind type not specified in header")
+    kind_type = container[counter]
+    counter += 1
+    if validate and kind_type != 1:
+        raise InvalidEof("Invalid kind type")
+
+    # Get the 2 bytes type_size
+    if validate and len(container) < counter + 2:
+        raise InvalidEof("Type size not specified in header")
+    type_size = Uint.from_be_bytes(container[counter : counter + 2])
+    counter += 2
+    if validate and (type_size < 4 or type_size > 4096 or type_size % 4 != 0):
+        raise InvalidEof("Invalid type size")
+    num_types = type_size // 4
+
+    # Get the 1 byte kind_code
+    if validate and len(container) < counter + 1:
+        raise InvalidEof("Kind code not specified in header")
+    kind_code = container[counter]
+    counter += 1
+    if validate and kind_code != 2:
```

Comment:
Is it possible to disentangle validation from the other logic? Perhaps in two functions, maybe with a dataclass to store intermediate values?

Just seems like this function is a bit complex.